tap_snapchat_ads/streams.py

```diff
@@ -17,758 +17,1419 @@
+        else:
+            key = bookmark_field
+
+        return (
+            state
+            .get('bookmarks', {})
+            .get(stream, {})
+            .get(key, default)
+        )
+
+    # To write new bookmark in output
+    def write_bookmark(self, state, stream, value, bookmark_field=None, parent=None, parent_id=None):
+        if parent and parent_id:
+            key = '{}(parent_{}_id:{})'.format(bookmark_field, parent, parent_id)
+        else:
+            key = bookmark_field
+        if 'bookmarks' not in state:
+            state['bookmarks'] = {}
+        if stream not in state['bookmarks']:
+            state['bookmarks'][stream] = {}
+
+        state['bookmarks'][stream][key] = value
+        LOGGER.info('Write state for Stream: {}, {} ID: {}, value: {}'.format(
+            stream, parent, parent_id, value))
+        singer.write_state(state)
+
+    # To transform string to datetime 
+    def transform_datetime(self, this_dttm):
+        import pdb; pdb.set_trace()
+        with Transformer() as transformer:
+            new_dttm = transformer._transform_datetime(this_dttm)
+        return new_dttm
+
+    # To process records
+    def process_records(self, catalog, #pylint: disable=too-many-branches
+                        stream_name,
+                        records,
+                        time_extracted,
+                        bookmark_field=None,
+                        max_bookmark_value=None,
+                        last_datetime=None):
+        stream = catalog.get_stream(stream_name)
+        schema = stream.schema.to_dict()
+        stream_metadata = metadata.to_map(stream.metadata)
+
+        with metrics.record_counter(stream_name) as counter:
+            for record in records:
+                # Transform record for Singer.io
+                with Transformer() as transformer:
+                    transformed_record = transformer.transform(
+                        record,
+                        schema,
+                        stream_metadata)
+
+                    # Reset max_bookmark_value to new value if higher
+                    if bookmark_field and (bookmark_field in transformed_record):
+                        bookmark_date = transformed_record.get(bookmark_field)
+                        bookmark_dttm = strptime_to_utc(bookmark_date)
+                        last_dttm = strptime_to_utc(last_datetime)
+
+                        if not max_bookmark_value:
+                            max_bookmark_value = last_datetime
+
+                        max_bookmark_dttm = strptime_to_utc(max_bookmark_value)
+
+                        if bookmark_dttm > max_bookmark_dttm:
+                            max_bookmark_value = strftime(bookmark_dttm)
+
+                        # Keep only records whose bookmark is after the last_datetime
+                        if bookmark_dttm >= last_dttm:
+                            # LOGGER.info('record1: {}'.format(record)) # TESTING, comment out
```

Comment:
These kinds of comments are present in multiple places. Remove those comments.

Reply:
Removed unnecessary comments.